Linking with x86_64-w64-mingw32 fails

Linking with x86_64-w64-mingw32 fails (works with i686-w64-mingw32):
Any clue what it might be ?

```
[ 50%] Linking CXX executable t_tbb.exe
CMakeFiles/t_tbb.dir/objects.a(t_tbb.cxx.obj): In function `tbb::interface9::internal::allocate_sibling(tbb::task*, tbb::task**, unsigned long long, unsigned long long)':
/usr/x86_64-w64-mingw32/include/tbb/parallel_reduce.h:175: undefined reference to `tbb::internal::allocate_continuation_proxy::allocate(unsigned long long) const'
/usr/x86_64-w64-mingw32/include/tbb/parallel_reduce.h:178: undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long long) const'
/usr/x86_64-w64-mingw32/include/tbb/parallel_reduce.h:175: undefined reference to `tbb::internal::allocate_continuation_proxy::allocate(unsigned long long) const'
/usr/x86_64-w64-mingw32/include/tbb/parallel_reduce.h:178: undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long long) const'
CMakeFiles/t_tbb.dir/objects.a(t_tbb.cxx.obj): In function `operator new(unsigned long long, tbb::internal::allocate_root_with_context_proxy const&)':
/usr/x86_64-w64-mingw32/include/tbb/task.h:1005: undefined reference to `tbb::internal::allocate_root_with_context_proxy::allocate(unsigned long long) const'
CMakeFiles/t_tbb.dir/objects.a(t_tbb.cxx.obj): In function `tbb::task_scheduler_init::task_scheduler_init(int, unsigned long long)':
/usr/x86_64-w64-mingw32/include/tbb/task_scheduler_init.h:118: undefined reference to `tbb::task_scheduler_init::initialize(int, unsigned long long)'
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/t_tbb.dir/build.make:99: t_tbb.exe] Error 1
make[2]: *** [CMakeFiles/Makefile2:810: CMakeFiles/t_tbb.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:822: CMakeFiles/t_tbb.dir/rule] Error 2
make: *** [Makefile:433: t_tbb] Error 2
```

Here the example I used:
```
#include <iostream>
#include <cstdlib>
#include <tbb/parallel_reduce.h>
#include <tbb/task_scheduler_init.h>
#include <tbb/blocked_range.h>
#include <tbb/partitioner.h>

static long num_rects = 2e9;

class MyPi {
  double *const my_rects;

public:
  double partialHeight;
  
  MyPi(double * const width) : my_rects(width), partialHeight(0) {}
 
  MyPi(MyPi & x, tbb::split) : my_rects(x.my_rects), partialHeight(0) {}
 
  void operator()(const tbb::blocked_range<size_t> &range) {
    double rectangleWidth = *my_rects;
    for (size_t i=range.begin(); i!=range.end(); ++i) {
      double x = (i + 0.5) * rectangleWidth;
      partialHeight += 4.0/(1. + x*x);
    }
  }
  // join threds and counters
  void join(const MyPi &other) {  
    partialHeight+=other.partialHeight;  
  }


};

int main(int argc, char *argv[])
{
  double area;
  double width = 1./(double)num_rects;
  MyPi my_block((double * const)&width);                    // functor for parallel reduce
  tbb::task_scheduler_init init;        // initiallize TBB task scheduler

  // parallel MC computation of pi 
  tbb::parallel_reduce(tbb::blocked_range<size_t>(0, num_rects), my_block, tbb::auto_partitioner());
  area = my_block.partialHeight * width;
  // print result
  std::cout << "pi = " << area << std::endl;
  return EXIT_SUCCESS;
}
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,15 +1,6 @@
 language: cpp
 sudo: false
-addons:
-  apt:
-    sources:
-      - ubuntu-toolchain-r-test
-    packages:
-      - g++-4.8
-      - g++-4.8-multilib
-      - g++-multilib
-      - binutils-mingw-w64-x86-64
-      - gcc-mingw-w64-x86-64
+
 matrix:
   include:
     - os: linux
@@ -38,8 +29,17 @@ matrix:
       addons:
         apt:
           packages:
-            - gcc-mingw-w64-base
-            - binutils-mingw-w64-x86-64
-            - gcc-mingw-w64-x86-64
             - g++-mingw-w64-x86-64
-            - gcc-mingw-w64
+            - gcc-mingw-w64-x86-64
+            - binutils-mingw-w64-x86-64
+    - os: linux
+      compiler: i686-w64-mingw32-g++
+      script:
+        - cmake -DCMAKE_CXX_COMPILER=i686-w64-mingw32-g++ -DCMAKE_TOOLCHAIN_FILE=build/mingw_cross_toolchain.cmake -DGNU_HOST=i686-w64-mingw32 .
+        - make -j 2
+      addons:
+        apt:
+          packages:
+            - g++-mingw-w64-i686
+            - gcc-mingw-w64-i686
+            - binutils-mingw-w64-i686

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -57,37 +57,35 @@ if (APPLE)
 endif()
 
 if (UNIX)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -DUSE_PTHREAD")
-  if(NOT CMAKE_CXX_FLAGS MATCHES "-mno-rtm")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mrtm")
-  endif()
+  set(CMAKE_CXX_FLAGS "-std=c++11 -DUSE_PTHREAD ${CMAKE_CXX_FLAGS}")
+  set(CMAKE_CXX_FLAGS "-mrtm ${CMAKE_CXX_FLAGS}")
 elseif(WIN32)
   if (MSVC)
     cmake_minimum_required (VERSION 3.1)
     enable_language(ASM_MASM)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /GS- /Zc:wchar_t /Zc:forScope /DUSE_WINTHREAD")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /D_CRT_SECURE_NO_DEPRECATE /D_WIN32_WINNT=0x0600")
+    set(CMAKE_CXX_FLAGS "/GS- /Zc:wchar_t /Zc:forScope /DUSE_WINTHREAD ${CMAKE_CXX_FLAGS}")
+    set(CMAKE_CXX_FLAGS "/D_CRT_SECURE_NO_DEPRECATE /D_WIN32_WINNT=0x0600 ${CMAKE_CXX_FLAGS}")
     check_cxx_compiler_flag ("/volatile:iso" SUPPORTS_VOLATILE_FLAG)
     if (SUPPORTS_VOLATILE_FLAG)
-      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /volatile:iso")
+      set(CMAKE_CXX_FLAGS "/volatile:iso ${CMAKE_CXX_FLAGS}")
     endif ()
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4267 /wd4800 /wd4146 /wd4244 /wd4018")
+    set(CMAKE_CXX_FLAGS "/wd4267 /wd4800 /wd4146 /wd4244 /wd4018 ${CMAKE_CXX_FLAGS}")
 
     if (CMAKE_SIZEOF_VOID_P EQUAL 8)
       list(APPEND tbb_src src/tbb/intel64-masm/atomic_support.asm
         src/tbb/intel64-masm/itsx.asm src/tbb/intel64-masm/intel64_misc.asm)
       list(APPEND tbbmalloc_src src/tbb/intel64-masm/atomic_support.asm)
-      set(CMAKE_ASM_MASM_FLAGS "${CMAKE_ASM_MASM_FLAGS} /DEM64T=1")
+      set(CMAKE_ASM_MASM_FLAGS "/DEM64T=1 ${CMAKE_ASM_MASM_FLAGS}")
     else()
       list(APPEND tbb_src src/tbb/ia32-masm/atomic_support.asm
         src/tbb/ia32-masm/itsx.asm src/tbb/ia32-masm/lock_byte.asm)
       # Enable SAFESEH feature for assembly (x86 builds only).
-      set(CMAKE_ASM_MASM_FLAGS "${CMAKE_ASM_MASM_FLAGS} /safeseh")
+      set(CMAKE_ASM_MASM_FLAGS "/safeseh ${CMAKE_ASM_MASM_FLAGS}")
     endif()
   elseif (MINGW)
     add_definitions(-DUSE_WINTHREAD)
     add_definitions(-D_WIN32_WINNT=0x0502)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mthreads")
+    set(CMAKE_CXX_FLAGS "-mthreads ${CMAKE_CXX_FLAGS}")
   endif ()
 endif()
 
@@ -102,17 +100,24 @@ endif ()
 # Linker export definitions
 if (APPLE)
   set (ARCH_PREFIX "mac")
-elseif(MSVC)
+elseif(WIN32)
   set (ARCH_PREFIX "win")
 else ()
-  # MinGW uses gcc def format
   set (ARCH_PREFIX "lin")
 endif ()
+
 if (CMAKE_SIZEOF_VOID_P EQUAL 8)
   set(ARCH_PREFIX "${ARCH_PREFIX}64")
 else()
   set(ARCH_PREFIX "${ARCH_PREFIX}32")
 endif()
+
+if (MINGW)
+  set (ARCH_PREFIX "${ARCH_PREFIX}-gcc")
+  # there's no win32-gcc-tbb-export.def, use lin32-tbb-export.def
+  execute_process (COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/src/tbb/lin32-tbb-export.def ${CMAKE_CURRENT_SOURCE_DIR}/src/tbb/win32-gcc-tbb-export.def)
+endif()
+
 if (MSVC)
   add_custom_command(OUTPUT tbb.def
     COMMAND ${CMAKE_CXX_COMPILER} /TC /EP ${CMAKE_CURRENT_SOURCE_DIR}/src/tbb/${ARCH_PREFIX}-tbb-export.def  -I ${CMAKE_CURRENT_SOURCE_DIR}/include > tbb.def

